business/src/main/java/org/mskcc/cbio/portal/service/ApiService.java

```diff
@@ -341,33 +326,107 @@ else if(type.equals("frequency"))
+	}
+
 	@Transactional
-	public List<Serializable> getGeneticProfileData(List<String> geneticProfileStableIds, List<String> hugoGeneSymbols,
-													List<String> sampleStableIds, String sampleListStableId) {
+	public List<DBProfileData> getGeneticProfileDataBySampleList(List<String> genetic_profile_ids, List<String> genes, String sample_list_id) {
+		return getGeneticProfileData(genetic_profile_ids, genes, null, sample_list_id);
+	}
 
-		List<DBGeneticProfile> profiles = getGeneticProfiles(geneticProfileStableIds);
-		List<String> mutationProfiles = new ArrayList<>();
-		List<String> nonMutationProfiles = new ArrayList<>();
-		for (DBGeneticProfile profile : profiles) {
-			if ("MUTATION_EXTENDED".equals(profile.genetic_alteration_type)) {
-				mutationProfiles.add(profile.id);
+	@Transactional
+	public List<DBProfileData> getGeneticProfileDataBySample(List<String> genetic_profile_ids, List<String> genes, List<String> sample_ids) {
+		return getGeneticProfileData(genetic_profile_ids, genes, sample_ids, null);
+	}
+
+	@Transactional
+	public List<DBProfileData> getGeneticProfileData(List<String> genetic_profile_ids, List<String> genes, List<String> sample_ids, String sample_list_id) {
+		List<DBGeneticProfile> profiles = getGeneticProfiles(genetic_profile_ids);
+		List<String> mutation_profiles = new ArrayList<>();
+		List<String> non_mutation_profiles = new ArrayList<>();
+		for (DBGeneticProfile p : profiles) {
+			if (p.genetic_alteration_type.equals("MUTATION_EXTENDED")) {
+				mutation_profiles.add(p.id);
 			} else {
-				nonMutationProfiles.add(profile.id);
+				non_mutation_profiles.add(p.id);
 			}
 		}
-
-		List<Serializable> result = new ArrayList<>();
-
-		if (!mutationProfiles.isEmpty()) {
-			result.addAll(addSampleListIdToMutationList(mutationService.getMutationsDetailed(mutationProfiles, hugoGeneSymbols, sampleStableIds, sampleListStableId), sampleListStableId));
+		List<DBProfileData> ret = new ArrayList<>();
+		if (!mutation_profiles.isEmpty()) {
+			List<DBMutationData> to_add;
+			if (sample_ids == null && sample_list_id == null) {
+				to_add = profileDataMapper.getMutationData(mutation_profiles, genes);
+			} else if (sample_list_id == null) {
+				to_add = profileDataMapper.getMutationDataBySample(mutation_profiles, genes, sample_ids);
+			} else {
+				to_add = profileDataMapper.getMutationDataBySampleList(mutation_profiles, genes, sample_list_id);
+			}
+			ret.addAll(to_add);
 		}
-		if (!nonMutationProfiles.isEmpty()) {
-			result.addAll(getNonMutationGeneticProfileData(nonMutationProfiles, hugoGeneSymbols, sampleStableIds,
-					sampleListStableId));
+		if (!non_mutation_profiles.isEmpty()) {
+			List<DBGeneticAltRow> genetic_alt_rows = profileDataMapper.getGeneticAlterationRow(non_mutation_profiles, genes);
+			List<DBProfileDataCaseList> ordered_sample_lists = profileDataMapper.getProfileCaseLists(non_mutation_profiles);
+
+			Set<String> desired_samples = new HashSet<>();
```

Comment:
i'd recommend breaking down the rest of the method into more fine-grained methods, not so much for reusability but to make the intent clearer (comments would do too but I prefer re-using the language's modularization constructs whenever possible)